os/fs/smartfs/smartfs_utils.c

```diff
@@ -1317,6 +1403,130 @@ int smartfs_createentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector,
+ *   sector - pointer to return the sector number allocated.
+ *   type - to set the type(file/dir) of the new sector allocated.
+ *   sf - pointer to smartfs ofile structure in case a new file is being
+ *     opened.
+ *
+ * Returned Values:
+ *   OK - on successful allocation of a new sector.
+ *   ERROR - apprpriate error code otherwise.
+ *
+ ****************************************************************************/
+
+int smartfs_alloc_firstsector(struct smartfs_mountpt_s *fs, uint16_t *sector, uint16_t type, struct smartfs_ofile_s *sf)
+{
+	int ret;
+	struct smart_read_write_s readwrite;
+	uint16_t alloc_sector;
+	uint8_t chainheader_type;
+#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+	struct smartfs_chain_header_s *chainheader;
+#endif
+
+	ret = FS_IOCTL(fs, BIOC_ALLOCSECT, 0xFFFF);
+	if (ret < 0) {
+		return ret;
+	}
+
+	alloc_sector = (uint16_t)ret;
+
+	/* Set the newly allocated sector's type (file or dir) */
+
+#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+	if (sf) {
+		/* Using sector buffer and we have an open file context. Just update
+		 * the sector buffer in the open file context.
+		 */
+		memset(sf->buffer, CONFIG_SMARTFS_ERASEDSTATE, fs->fs_llformat.availbytes);
+		chainheader = (struct smartfs_chain_header_s *)sf->buffer;
+		chainheader->type = SMARTFS_SECTOR_TYPE_FILE;
+		sf->bflags = SMARTFS_BFLAG_DIRTY | SMARTFS_BFLAG_NEWALLOC;
+	} else
+#endif
+	{
+		if ((type & SMARTFS_DIRENT_TYPE) == SMARTFS_DIRENT_TYPE_DIR) {
+			chainheader_type = SMARTFS_SECTOR_TYPE_DIR;
+		} else {
+			chainheader_type = SMARTFS_SECTOR_TYPE_FILE;
+		}
+
+		smartfs_setbuffer(&readwrite, alloc_sector, offsetof(struct smartfs_chain_header_s, type), 1, (uint8_t *)&chainheader_type);
+		ret = FS_IOCTL(fs, BIOC_WRITESECT, (unsigned long)&readwrite);
+		if (ret < 0) {
+			fdbg("Error %d setting new sector type for sector %d\n", ret, alloc_sector);
+			return ret;
+		}
+	}
+
+	*sector = alloc_sector;
+	return OK;
+}
+
+
+/****************************************************************************
+ * Name: smartfs_invalidateentry
+ *
+ * Description: Invalidates the smartfs entry passed as a parameter.
+ *
+ * Input Parameters:
+ *  fs - pointer to smartfs mountpoint structure.
+ *  oldentry - entry to be marked invalid.
+ *
+ * Returned Values:
+ *  OK - on sucessfully invalidating the entry.
+ *  ERROR - appropriate error code otherwise
+ *
+ ****************************************************************************/
+
+int smartfs_invalidateentry(struct smartfs_mountpt_s *fs, struct smartfs_entry_s oldentry)
+{
+	int ret;
+	uint8_t *tmp_pntr = NULL;
+	uint16_t tmp_flag = 0;
+	struct smart_read_write_s readwrite;
+
+	smartfs_setbuffer(&readwrite, oldentry.dsector, 0, fs->fs_llformat.availbytes, (uint8_t *)fs->fs_rwbuffer);
+	ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+	if (ret < 0) {
+		fdbg("Error %d reading sector %d data\n", ret, oldentry.dsector);
+		return ret;
+	}
+
+	// bug fix for rename
```

Comment:
Actually another guy fixed this about 5-6 years ago but now I can't remember it.
but now I think this is useless, let me hear your opinion.

Reply:
Let's align this code with all other functions.
We'll use a `smartfs_entry_header_s` structure for making changes to the entry flags.